src/Sylius/Behat/PageObjectExtension/PageObject/Page.php

```diff
@@ -0,0 +1,282 @@
+        $this->parameters = $parameters;
+
+        $this->document = new DocumentElement($session);
+    }
+
+    /**
+     * @param array $urlParameters
+     *
+     * @return Page
+     */
+    public function open(array $urlParameters = [])
+    {
+        $url = $this->getUrl($urlParameters);
+
+        $this->getDriver()->visit($url);
+
+        $this->verify($urlParameters);
+
+        return $this;
+    }
+
+    /**
+     * @param array $urlParameters
+     */
+    public function verify(array $urlParameters)
+    {
+        $this->verifyResponse();
+        $this->verifyUrl($urlParameters);
+        $this->verifyPage();
+    }
+
+    /**
+     * @param array $urlParameters
+     *
+     * @return bool
+     */
+    public function isOpen(array $urlParameters = [])
+    {
+        try {
+            $this->verify($urlParameters);
+        } catch (\Exception $e) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * @return string
+     */
+    abstract protected function getPath();
+
+    /**
+     * @return string
+     */
+    protected function getName()
+    {
+        return preg_replace('/^.*\\\(.*?)$/', '$1', get_called_class());
+    }
+
+    /**
+     * @param array $urlParameters
+     *
+     * @return string
+     */
+    protected function getUrl(array $urlParameters = [])
+    {
+        return $this->makeSurePathIsAbsolute($this->unmaskUrl($urlParameters));
+    }
+
+    /**
+     * @throws UnexpectedPageException
+     */
+    protected function verifyResponse()
+    {
+        try {
+            $statusCode = $this->getDriver()->getStatusCode();
+
+            if ($this->isErrorResponse($statusCode)) {
+                $currentUrl = $this->getDriver()->getCurrentUrl();
+                $message = sprintf('Could not open the page: "%s". Received an error status code: %s', $currentUrl, $statusCode);
+
+                throw new UnexpectedPageException($message);
+            }
+        } catch (DriverException $exception) {
+            // ignore drivers which cannot check the response status code
+        }
+    }
+
+    /**
+     * Overload to verify if the current url matches the expected one. Throw an exception otherwise.
+     *
+     * @param array $urlParameters
+     */
+    protected function verifyUrl(array $urlParameters = [])
+    {
+        if ($this->getDriver()->getCurrentUrl() !== $this->getUrl($urlParameters)) {
+            throw new UnexpectedPageException(sprintf('Expected to be on "%s" but found "%s" instead', $this->getUrl($urlParameters), $this->getDriver()->getCurrentUrl()));
+        }
+    }
+
+    /**
+     * Overload to verify if we're on an expected page. Throw an exception otherwise.
+     */
+    protected function verifyPage()
+    {
+    }
+
+    /**
+     * @param string $statusCode
+     *
+     * @return bool
+     */
+    protected function isErrorResponse($statusCode)
+    {
+        return in_array(substr($statusCode, 0, 1), ['4', '5'], true);
```

Comment:
The `$statusCode` is `int` so we can check `400 <= $statusCode && $statusCode < 600`